### PR TITLE
Simulation.h: fix type-matching and const errors in templates.

### PR DESCRIPTION
**Pull request by rpoyner-tri**
* simulate(5-arg): don't modify the passed options struct.
   * restore const to options arg, to get sim(4) a fighting chance.
 * simulate(4-arg): fix a typo that defeats template resolution.

### CODE REVIEW

**Comment by rpoyner-tri** [comment]
@jwnimmer-tri: please review this PR.

File: drake/systems/Simulation.h [hunk 1]
```diff
@@ -103,8 +103,9 @@ inline bool handle_realtime_factor(const TimePoint& wall_clock_start_time,
 template <typename System>
 void simulate(const System& sys, double ti, double tf,
               const typename System::template StateVector<double>& xi,
-              SimulationOptions& options) {
-  if (options.realtime_factor < 0.0) options.realtime_factor = 0.0;
+              const SimulationOptions& options) {
+  SimulationOptions my_options = options;
+  if (my_options.realtime_factor < 0.0) my_options.realtime_factor = 0.0;
```
Comment (jwnimmer-tri) [inline drake/systems/Simulation.h]: Did you investigate whether any calling code relies on this broken output-writeback behavior?

File: drake/systems/Simulation.h [hunk 2]
```diff
@@ -103,8 +103,9 @@ inline bool handle_realtime_factor(const TimePoint& wall_clock_start_time,
 template <typename System>
 void simulate(const System& sys, double ti, double tf,
               const typename System::template StateVector<double>& xi,
-              SimulationOptions& options) {
-  if (options.realtime_factor < 0.0) options.realtime_factor = 0.0;
+              const SimulationOptions& options) {
+  SimulationOptions my_options = options;
```
Comment (jwnimmer-tri) [inline drake/systems/Simulation.h]: Consider: rename the arg to be input_options and keep the local as "options"... possible hazard that anyone editing this method would accidentally use "options.something" and not realize that my_options was the preferred spelling within the method.

File: drake/systems/Simulation.h [hunk 3]
```diff
@@ -162,7 +163,7 @@ void simulate(const System& sys, double ti, double tf,
  */
 template <typename System>
 void simulate(const System& sys, double t0, double tf,
-              const typename System::template StateVectorType<double>& x0) {
+              const typename System::template StateVector<double>& x0) {
```
Comment (jwnimmer-tri) [inline drake/systems/Simulation.h]: This seems correct, but I have no idea why the old code even compiled at all.  Was nobody using it?

**Comment by jwnimmer-tri** [comment]
Comments in notes.  LGTM.

**Comment by rpoyner-tri** [comment]
@jwnimmer-tri: i did not investigate reliance on brokenness further than "no tests failed". @david-german-tri assures that broader reform is imminent, so I didn't dig any deeper.

As for sim(arity-4), it was clearly just dead code before I resuscitated it.

File: drake/systems/Simulation.h (hunk 2 above)
Comment (david-german-tri) [inline drake/systems/Simulation.h]: Could you avoid that ambiguity by not copying the options at all, and just applying the max at the point of use?  `</drive-by-review>`

Comment (jwnimmer-tri) [inline drake/systems/Simulation.h]: That sounds fine to me, too.

**Comment by rpoyner-tri** [comment]
I'll incorporate @david-german-tri 's suggestion and send it through again.

File: drake/systems/Simulation.h (hunk 1 above)
Comment (jwnimmer-tri) [inline drake/systems/Simulation.h]: FWIW, I did the grep for this, and didn't see any code that touched realtime_factor after simulate() or runLCM() is called.  (I did find that runLCM re-implements the "negative means 1.0" rule, but _shrug_.)

**Comment by rpoyner-tri** [comment]
revised patch is up. @jwnimmer-tri , @david-german-tri , please have a look. Thanks!

**Comment by jwnimmer-tri** [comment]
LGTM.

**Comment by david-german-tri** [comment]
LGTM as well

**Comment by jwnimmer-tri** [comment]
@david-german-tri Hmm.  I think we are going to need to close (and tag ang) and reopen this, to get drake005 coverage?

**Comment by rpoyner-tri** [comment]
didn't drake0005 already run?

**Comment by jwnimmer-tri** [comment]
Huh.  When I commented, it had not, but its there now.

Ah, I see.  More of us got added to the whitelist, so that the "approved for build servers" tag is not required.

**Comment by jwnimmer-tri** [comment]
Two reviews done, four CI reports green, bombs away!
